### PR TITLE
feat(state-transition): add deposits root to beacon state

### PR DESCRIPTION
**Pull request by abi87**
NOTE: this change is state breaking with respect to bartio and boonet

Introduce deposit validation in state.

- Include merkle root for all deposits in block.Eth1Data.DepositRoot
- recheck block.Eth1Data.DepositRoot against deposit root recalculated from local state, accounting for deposits coming in the block
- drop deposits pruning to preserve all deposits for the state verification

### CODE REVIEW

File: state-transition/core/state/statedb.go [hunk 1]
```diff
@@ -389,6 +389,11 @@ func (s *StateDB[BeaconStateMarshallableT, _]) GetMarshallable() (BeaconStateMar
 		return empty, err
 	}
 
+	deposits, err := s.GetDeposits()
```
Comment (calbera) [inline state-transition/core/state/statedb.go:392]: should return the deposits of the last processed block

File: state-transition/core/state_processor_staking.go [hunk 2]
```diff
@@ -60,6 +60,10 @@ func (sp *StateProcessor[
 			return err
 		}
 	}
+
+	// Add deposits to beaconState
+	st.SetDepositRoot(deposits)
```
Comment (calbera) [inline state-transition/core/state_processor_staking.go:65]: UpdateDepositRoot AND set block deposits in the state

File: consensus-types/types/state.go [hunk 3]
```diff
@@ -58,6 +58,9 @@ type BeaconState struct {
 	// Slashing
 	Slashings     []math.Gwei
 	TotalSlashing math.Gwei
+
+	// Deposits
+	DepositsRoot common.Root
```
Comment (sbudella-gco) [inline consensus-types/types/state.go:63]: This is not needed if we stick to `Eth1Data`

File: consensus-types/types/state.go [hunk 4]
```diff
@@ -97,6 +101,7 @@ func (st *BeaconState) New(
 		NextWithdrawalValidatorIndex: nextWithdrawalValidatorIndex,
 		Slashings:                    slashings,
 		TotalSlashing:                totalSlashing,
+		DepositsRoot:                 depositsRoot,
```
Comment (sbudella-gco) [inline consensus-types/types/state.go:104]: same as above.

File: consensus-types/types/state.go (hunk 3 above)
Comment (abi87) [reply consensus-types/types/state.go]: Dropped

File: consensus-types/types/state.go (hunk 4 above)
Comment (abi87) [reply consensus-types/types/state.go]: Dropped

**Review by calbera — COMMENTED** [review]
In theory, once changing the `maxItems` in `Deposits` HashTreeRoot this seems correct to me/

File: state-transition/core/core_test.go [hunk 5]
```diff
@@ -227,3 +227,10 @@ func buildNextBlock(
 		Body:          nextBlkBody,
 	}
 }
+
+// We verify DepositRoot in Eth1Data, but we don't really update
+// the store once we store deposits (it's not StateProcessor responsibility).
+// So we just keep using the dummyEth1Data most of the time.
+var dummyEth1Data = &types.Eth1Data{
```
Comment (sbudella-gco) [inline state-transition/core/core_test.go:234]: no longer needed

Comment (calbera) [reply state-transition/core/core_test.go:234]: addressed in https://github.com/berachain/beacon-kit/pull/2304

**Comment by shotes** [comment]
Since we are calling `GetDepositsByIndex` on the whole list of deposits, we may want to consider optimizing that function by pre-allocating [`deposits = []*ctypes.Deposit{}`](https://github.com/berachain/beacon-kit/blob/c08698ebbc62b5f6496ed400622978e10adbd2d2/storage/deposit/store.go#L83) to a length of `endIdx - startIndex` instead of appending a new deposit and causing a bunch of reallocation.

Which brings me to another question regarding that same function `GetDepositsByIndex` - why are we returning no error when a deposit index is not found?

**Comment by shotes** [comment]
The genesis deposits don't seem to be added to the `DepositStore` even though the `Eth1DepositIndex` gets incremented during genesis. This means that `GetDepositsByIndex(0, depositIndex)` would return an empty list, and all deposit roots would not include any of the previous deposits outside of the current new set.

**Review by calbera — APPROVED** [review]

File: primitives/constants/payload.go [hunk 6]
```diff
@@ -31,8 +31,12 @@ const (
 	// payload.
 	MaxTxsPerPayload uint64 = 1048576
 
-	// MaxDepositsPerBlock is the maximum number of deposits per block.
-	MaxDepositsPerBlock uint64 = 16
+	// DepositContractDepth is the depth of the deposit contract merkle tree.
+	DepositContractDepth uint64 = 32
```
Comment (abi87) [inline primitives/constants/payload.go:35]: q: @calbera does this require changes to the deposit contract?

**Pull request merged**